Compatible for VRCConstraint

kesera2 opened the pull request:
# プルリクエストの説明
このプルリクエストでは、ConstraintからVRCConstraintに対応する変更を追加しました。これにより、VRChat向けのConstraintコンポーネントが使用できるようになります。

# 環境
- Unity 2022.3.22f1
- VRChat SDK, Avatar 3.7.5

# 変更内容
ConstraintEditor.cs内で、さまざまな制約タイプに対してVRCConstraintを追加しました。
AimConstraint、LookAtConstraint、ParentConstraint、PositionConstraint、RotationConstraint、およびScaleConstraintに対して、適切なVRCバージョンを対応付けました。
各制約が正しく追加されるよう、関連するロジックを修正しました。

# 動機
VRChatでのアバター改変において、Unity Built-in ConstraintよりVRCConstraintを使用することが安定性や軽量化を踏まえて重要だと考えています。この変更により、ユーザーがより簡単にVRChat向けの制約を設定できるようになります。

# テスト
各VRCConstraintを持つアバターを「Apply Changes」より作成し、動作が期待通りであることを確認しました。各VRCConstraintが正しく機能することを確認しました。

この変更がプロジェクトに役立つことを願っています。ご確認のほどよろしくお願いいたします

kesera2 committed:
Compatible for VRCConstraint

kesera2 commented:
ささやかですがBOOTHにてDonationさせていただきました。

mika-f commented:
コントリビューションありがとうございます！しかしながら、 Constraint by Humanoid は現状 VRChat 専用のツールというわけではなく、 Unity 汎用のツールとして作成しております。そのため、  VRChat SDK が必須となってしまう現状のまま変更を受け入れるのは厳しいです。
VRChat SDK が読み込まれている場合 (特定のプリプロセッサシンボルが定義されている場合) のみ VRCConstraint に切り替わる、などといった形式だとマージすることが可能になるのですが、こちらいかがでしょうか？

kesera2 committed:
Add VRC_SDK_Base_2022 to asmdef

kesera2 committed:
Compatible for built-in and vrcsdk

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -9,8 +9,11 @@
 using System.Text.RegularExpressions;
 using UnityEditor;
 using UnityEngine;
+using UnityEngine.Animations;
```
built-in環境で使用するConstraintの名前空間です。

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -9,8 +9,11 @@
 using UnityEditor;
 using UnityEngine;
+using UnityEngine.Animations;
+#if VRC_SDK_Base_2022
```
プリプロセッサディレクティブによりbuilt-in環境でVRCSDKの名前空間がない事により発生するエラーを回避します。

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -235,19 +238,31 @@ private static void ConfigureRotationConstraintToGameObject(Animator src, Animat
                 return;
 
-            if (dstGameObject.GetComponent(GetTypeFromConstraint(type)) != null)
+#if VRC_SDK_Base_2022
```
可読性を維持するために条件付きコンパイルを複数多用するのを控えました。
差分だけ適用する場合以下の様になるかと思います。
```csharp
#if VRC_SDK_Base_2022
            if (dstGameObject.GetComponent(GetTypeFromVrcConstraint(type)) != null)
#else
            if (dstGameObject.GetComponent(GetTypeFromConstraint(type)) != null)
#endif
            {
                Debug.LogWarning($"The GameObject `{dstGameObject.name}` has been skipped because it already has {type.ToString()}.");
                return;
            }
#if VRC_SDK_Base_2022            
            var constraint = AddVrcConstraintToGameObject(dstGameObject, type);
            var source = new VRCConstraintSource { SourceTransform = srcGameObject.transform, Weight = 1.0f };
            constraint.Sources.Add(source);
            constraint.ActivateConstraint();
#else
            var constraint = AddConstraintToGameObject(dstGameObject, type);
            var source = new ConstraintSource { sourceTransform = srcGameObject.transform, weight = 1.0f };
            constraint.AddSource(source);
            constraint.constraintActive = true;
#endif
```

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -300,7 +315,61 @@ private static VRCConstraintBase AddConstraintToGameObject(GameObject obj, Const
             }
         }
+#else
+        private static Type GetTypeFromConstraint(Constraint constraint)
```
Built-in用のメソッドを復元しました。

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -300,7 +315,61 @@ private static VRCConstraintBase AddConstraintToGameObject(GameObject obj, Const
+            }
+        }
+
+        private static IConstraint AddConstraintToGameObject(GameObject obj, Constraint constraint)
```
Built-in用のメソッドを復元しました。

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -274,7 +289,7 @@ private static Type GetTypeFromConstraint(Constraint constraint)
         }
 
-        private static VRCConstraintBase AddConstraintToGameObject(GameObject obj, Constraint constraint)
+        private static VRCConstraintBase AddVrcConstraintToGameObject(GameObject obj, Constraint constraint)
```
メソッド名にVrcの語句をつけBuilt-inと区別しました。

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -235,19 +238,31 @@ private static void ConfigureRotationConstraintToGameObject(Animator src, Animat
-        private static Type GetTypeFromConstraint(Constraint constraint)
+        
+#if VRC_SDK_Base_2022
+        private static Type GetTypeFromVrcConstraint(Constraint constraint)
```
メソッド名にVrcの語句をつけBuilt-inと区別しました。

kesera2 commented:
ご確認ありがとうございます。VRChat環境だけではなくUnity 汎用のツールとして利用できる形に再度修正を加えました。

> VRChat SDK が読み込まれている場合 (特定のプリプロセッサシンボルが定義されている場合) のみ VRCConstraint に切り替わる、などといった形式だとマージすることが可能になるのですが、こちらいかがでしょうか？


## 言葉の定義
- Built-inまたはUnity Built-in環境 ... VRCSDKがパッケージとしてインポートされていないUnityプロジェクトのことを指します。
- VRCSDK環境 ... VRCSDKがパッケージとしてインポートされているUnityプロジェクトのことを指します。

# 変更内容
## Assembly Definitionの変更
NatsunekoLaboratory.ConstraintByHumanoid.asmdefを変更し以下の設定を行いました。

- Version Definesへcom.vrchat.baseをVRC_SDK_Base_2022として追加。ExpressionをUnity2022におけるVRCSDKの最初のセマンティックバージョンとして[3.5.0](https://creators.vrchat.com/releases/release-3-5-0)以上を指定。
- アバター用プロジェクトとワールド用プロジェクト両方での利用が見込まれるため、VRCSDKの対象のパッケージをcom.vrchat.baseとさせていただきました。
Unityのバージョンの選定理由として、VRChatはUnity2019のサポートを終了しているため、Unity2022を採用した形となります。
![image](https://github.com/user-attachments/assets/79cb1aff-568c-41c6-b8d2-6932497a1e76)

## コードの変更の概要
1. 上記で用意したプリプロセッサシンボルを使用し、Unity Built-in環境とVRCSDK環境それぞれにおいて適切なConstraintを設定できるように修正しました。
2. Built-in環境用に加え、VRChat環境用にGetTypeFromVrcConstraintとAddVrcConstraintToGameObjectのようにVrcの語句を追加し、プリプロセッサシンボルで使用するメソッドを切り替えるようにしました。
3. プリプロセッサシンボルを用いて、名前空間からインポートする型を切り替えるようにしました。

# テスト
## 使用アセット
- [sakura_ShowIcon/UnityのHierarchyにComponentアイコンを表示するやつ](https://tbbsakura.booth.pm/items/4477191)
- [オリジナル3Dモデル『コロネ』](https://booth.pm/ja/items/5776079)
- [オリジナル3Dモデル『クララ』](https://emolab.booth.pm/items/5484178)
- [【無料】lilToon](https://booth.pm/ja/items/3087170)

## Unity 2022.3.22f1 VRCSDK 3.7.5 プロジェクトにてApply Changesによる動作確認
1. VCC(ALCOM)より上記環境の新規プロジェクト作成
4. テスト用のアバター(ClaraとCornet)とシェーダー(lilToon)をインポート
5. ClaraからCornetに対してEnumに用意されているすべてのConstraintを順番に適用
6. VRCConstraintとして設定が正しく反映されているかを確認
![image](https://github.com/user-attachments/assets/b6d49a54-ed24-4caa-bb7b-f3a4aa942eba)

## Unity 2022.3.22f1 汎用3DプロジェクトにてApply Changesによる動作確認
1. Unity Hubより上記環境の新規プロジェクト作成
2. テスト用のアバター(ClaraとCornet)とシェーダー(lilToon)をインポート
3. ClaraからCornetに対してEnumに用意されているすべてのConstraintを順番に適用
4. Built-in Constraintとして設定が正しく反映されているかを確認
![image](https://github.com/user-attachments/assets/923d64eb-841a-421d-8b97-21b867198b8c)

# 参考
Detecting the VRChat SDK
https://creators.vrchat.com/sdk/detecting-vrcsdk/
条件付きコンパイルのdefineとしてVRC_SDK_VRCSDK3の利用を当初検討しましたが、廃止予定のため採用を見送りました。

条件付きコンパイル
https://docs.unity3d.com/ja/2022.3/Manual/PlatformDependentCompilation.html

kesera2 reviewed (commented):
変更内容について詳細をコメントに追加しました。

mika-f commented:
Approve した後に気がついたのですが、 [現状 World 用プロジェクトでは VRCConstraint が使えるか微妙そうな気がする](https://creators.vrchat.com/worlds/whitelisted-world-components#unity-components-1)ので、せっかく SDK3 Base でやってもらったのに申し訳ないのですが、可能でしたら SDK3 Avatars のときのみ適用、 SDK3 World は未適用、というのは可能でしょうか。

kesera2 committed:
Excluded the com.vrchat.worlds from version defines

Reason: VRC constraint is not supported by VRC World SDK. Only VRC Avatar SDK is supported now.

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -10,7 +10,7 @@
 using UnityEngine;
 using UnityEngine.Animations;
-#if VRC_SDK_Base_2022
+#if VRC_SDK_AVATARS_2022
```
Version Definesをcom.vrchat.baseからcom.vrchat.avatarsに変更し、WorldSDKを除外しました。

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -238,7 +238,7 @@ private static void ConfigureRotationConstraintToGameObject(Animator src, Animat
                 return;
 
-#if VRC_SDK_Base_2022
+#if VRC_SDK_AVATARS_2022
```
Version Definesをcom.vrchat.baseからcom.vrchat.avatarsに変更し、WorldSDKを除外しました。

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/Editor/ConstraintEditor.cs:
```diff
@@ -261,7 +261,7 @@ private static void ConfigureRotationConstraintToGameObject(Animator src, Animat
         }
         
-#if VRC_SDK_Base_2022
+#if VRC_SDK_AVATARS_2022
```
Version Definesをcom.vrchat.baseからcom.vrchat.avatarsに変更し、WorldSDKを除外しました。

kesera2 commented:
@mika-f 
せっかくAprroveしてもらったのに申し訳ないです。
Version Definesをcom.vrchat.baseからcom.vrchat.avatarsに変更し、WorldSDKを除外しました。

## Unity 2022.3.22f1 World用プロジェクトにてApply Changesによる動作確認
1. VCC(ALCOM)より上記環境の新規プロジェクト作成
2. テスト用のアバター(ClaraとCornet)とシェーダー(lilToon)をインポート
3. ClaraからCornetに対してEnumに用意されているすべてのConstraintを順番に適用
4. "**Built-in Constraint**"として設定が正しく反映されているかを確認
![image](https://github.com/user-attachments/assets/7a662153-bccc-4b8c-a572-9f1132dcab2e)

kesera2 commented on Assets/NatsunekoLaboratory/ConstraintByHumanoid/NatsunekoLaboratory.ConstraintByHumanoid.asmdef:
```diff
@@ -10,5 +10,13 @@
+        {
+            "name": "com.vrchat.avatars",
+            "expression": "3.5.0",
+            "define": "VRC_SDK_AVATARS_2022"
```
Version Definesをcom.vrchat.baseからcom.vrchat.avatarsに変更し、WorldSDKを除外し、
defineをVRC_SDK_AVATARS_2022としました。

mika-f commented:
良さそうなのでマージしました！

kesera2 commented:
@mika-f ありがとうございます！フィードバックやサポートに感謝します。

今後ともよろしくお願いいたします。
